portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java

```diff
@@ -0,0 +1,157 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.ListIterator;
+
+/**
+ * Models an implementation of the TimeRangeGroup model using LinkedList. This is a slightly
+ * modified version of the ArrayList implementation.
+ */
+public class LinkedListTimeRangeGroup extends AbstractListTimeRangeGroup implements TimeRangeGroup {
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    // If the original allTimeRanges list is empty,
+    // this is the first time we add anything to the list,
+    // so simply add the new time range to the list and return.
+    if (allTimeRanges.isEmpty()) {
+      allTimeRanges.add(timeRange);
+      return;
+    }
+    ListIterator<TimeRange> iterator = allTimeRanges.listIterator();
+
+    // This variable represents the latest time range previously exmamined.
+    // Initially, this variable points to the time range we want to add.
+    // As the for loop iterates through all the current time ranges already existing
+    // in allTimeRanges, this lastExaminedTimeRange variable is the time range with which
+    // any existing time range merges with, if necessary.
+    TimeRange lastExaminedTimeRange = timeRange;
+
+    // We need to start off with a currentRange inside the loop
+    TimeRange currentRange = iterator.next();
+
+    // The reason I used a true is because moving back and forth in the
+    // linked list can be confusing. Each case should be handled differently
+    // so using iterator.hasNext() and currentRange=iterator.next() in each
+    // iteration of this loop would make things harder to keep track of.
+    while (true) {
+
+      // If the current range is completely contained by the lastExaminedTimeRange,
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)
+          && !lastExaminedTimeRange.equals(currentRange)) {
+        iterator.remove();
+        if (iterator.hasNext()) {
+          currentRange = iterator.next();
+        }
+      }
+
+      // The case for when two time ranges need to be merged.
+      // Similar to the case above, it will eventually be replaced so we can
+      // remove currentRange
+      if (lastExaminedTimeRange.overlaps(currentRange)
```

Comment:
Can this be an else if? If so, then we won't need the ! contains check.

Reply:
I think this can't be an else if because the if statement that comes before this and the else if statement after this are not mutually exclusive.